.github/workflows/ci.yml

```diff
@@ -22,6 +23,21 @@ on:
   pull_request:
     branches:
       - main
+  # Support merge queues.
+  merge_group:
+    types:
+      - checks_requested
+  # Allow manual invocation – useful for debugging.
+  workflow_dispatch:
+
+# Cancel any previously-started but still active runs on the same branch.
+concurrency:
+  cancel-in-progress: true
+  group: ${{github.workflow}}-${{github.event.pull_request.number||github.ref}}
+
+# Declare default permissions as read-only.
```

Comment:
permissions for what?

Reply:
Permissions granted to workflow actions w.r.t. GITHUB_TOKEN.

C.f. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token